### PR TITLE
Fix python Ops proposal

### PR DESCRIPTION
**Pull request by philei-tt**
### Ticket
N/A

### Problem description
Resolve some issues I found in python ops proposal

### What's changed
- Fix multi output op dependency handling
- Fix handling of uninitialized gradients (fill them with zeros) (to match c++)
- Remove some weird code
- Remove branch pruning optimization
- Add more complicated tests: 
  - linear operation 
  - grouped_head_creation

### Checklist

N/A

### CODE REVIEW

**Comment by philei-tt** [comment]
Why if some of the output gradients not initialized, we should initialize it with zeros instead of failing. 

Take a look at `test_python_grouped_heads_creation_multi_output_nodes`

```
q, k, v = PythonGroupedHeadsCreationOp.apply(qs, kvs, num_heads, num_groups)
loss = ttml.ops.unary.mean(q)  # Only use q
loss.backward(retain_graph=False)
```

Operation has multiple outputs, but only one of them is used for loss calculation. We still want to be able to calculate backward for such case, but kvs.grad is going to be uninitialized, since it's not used in graph. Before it would fail, I argue that we should just init them with zero (dL/dkvs = 0 - they don't constribute to loss), I think it works similarly in c++. 

I think pytorch uses Nones for such cases, but then we will have to handle Nones in backward implementation, and I don't want that right now. We might change that in the future for memory/performance improvements
